### ISSUE TITLE
Regarding arbitrary mask

### ISSUE THREAD

**Issue by netw0rkf10w** [idiap/fast-transformers#22]
Hi,

I observe that only non-masked and causally-masked linear attentions have been implemented, and it's not clear to me why arbitrary mask can be difficult to implement.

Let `m` be the mask of the input sequence, where `m_i = 0` means the token `i` is masked out (the inverse of `m` is the same as the `src_key_padding_mask` argument in `torch.nn.Transformer`). Then Equation (5) in your paper becomes: 

<a href="https://www.codecogs.com/eqnedit.php?latex=V_i'&space;=&space;\frac{\phi(Q_i)^T\sum_{j=1}^Nm_j\phi(K_j)V_j^T}{\phi(Q_i)^T\sum_{j=1}^Nm_j\phi(K_j)}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?V_i'&space;=&space;\frac{\phi(Q_i)^T\sum_{j=1}^Nm_j\phi(K_j)V_j^T}{\phi(Q_i)^T\sum_{j=1}^Nm_j\phi(K_j)}" title="V_i' = \frac{\phi(Q_i)^T\sum_{j=1}^Nm_j\phi(K_j)V_j^T}{\phi(Q_i)^T\sum_{j=1}^Nm_j\phi(K_j)}" /></a>

Is that correct?

By the way, in your causal mask [implementation](https://github.com/idiap/fast-transformers/blob/master/fast_transformers/attention/causal_linear_attention.py), the mask doesn't seem to be applied anywhere (sorry if I've overlooked that).

Thank you in advance.

**Comment by angeloskath** [comment]
Hi,

The type of mask you have written above can indeed be implemented and a certain form of it is. The length mask for keys is implemented exactly that way. However, note that **this mask is fixed for all queries**. Hence it is not an arbitrary mask. An arbitrary mask would be represented by `m_ij` and that cannot be included the above computation.

Regarding the causal mask implementation, the mask is implemented using looping so no need to apply the mask with a multiplication or addition. You can check out the [cpu implementation](https://github.com/idiap/fast-transformers/blob/ea9cb3b1751f0b2f1e661b087f3dd3ec8a413ab0/fast_transformers/causal_product/causal_product_cpu.cpp#L114) which is simpler.

Cheers,
Angelos

**Comment by angeloskath** [comment]
Hi again,

I am closing the issue but if you have more questions or regarding the masking feel free to reopen it.

Cheers,
Angelos

**Comment by netw0rkf10w** [comment]
Hi @angeloskath. Thanks for the reply.
You are right that the 1D mask that I described above is not the most general one (which is supposed to be 3D: `(batch * heads, target_length, source_length)`), but for self-attention it should cover most of the standard applications (e.g. BERT in which the source sentence is randomly masked, your current implementation doesn't seem to cover this use case).

**Comment by angeloskath** [comment]
So you 'd want a generalization of the key_lengths mask where you can mask a key for all queries. Is that correct? I am just making sure that it's clear that a different mask per query is not possible.

**Comment by netw0rkf10w** [comment]
> So you 'd want a generalization of the key_lengths mask where you can mask a key for all queries. Is that correct?

Yes, because I think that's the case for e.g. BERT, isn't it?

**Comment by angeloskath** [comment]
I don't think it is but for sure this is a capability that should be possible. So, I will make another issue with the label enhancement and refer this one.

Cheers,
Angelos
